modules/product/src/Form/AddToCartForm.php

```diff
@@ -166,4 +182,253 @@ function submitForm(array &$form, FormStateInterface $form_state) {
     ]));
   }
 
+  /**
+   * Builds the variation selection form.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   * @param \Drupal\commerce_product\Entity\ProductVariationInterface[] $variations
+   *   An array of product variations.
+   */
+  protected function variationForm(array &$form, FormStateInterface $form_state, array $variations) {
+    // Resolve the default variation from current attribute values.
+    $this->resolveDefaultVariation($variations, $form_state);
+
+    // Get qualified attribute options and add them to the form.
+    foreach ($this->resolveQualifyingAttributes($variations) as $name => $data) {
+      $this->attributeElement($form, $data, $data['used_options']);
```

Comment:
If we're already passing $data, then there's no need to pass $data['used_options'] separately.